types/react-router/index.d.ts

```diff
@@ -162,7 +162,7 @@ export function useHistory<HistoryLocationState = H.LocationState>(): H.History<
 
 export function useLocation<S = H.LocationState>(): H.Location<S>;
 
-export function useParams<Params extends { [K in keyof Params]?: string } = {}>(): { [p in keyof Params]: keyof Params[p] extends undefined ? string | undefined : string  };
+export function useParams<Params extends { [K in keyof Params]?: string } = {}>(): { [K in keyof Params]: Params[K] };
```

Comment:
Ok let's go back a bit: Why do we need to map `{ optional?: string }` to `{ optional?: string | undefined }`? Would it make more sense to simplify this to 
```suggestion
export function useParams<Params extends { [K in keyof Params]?: string } = {}>(): Params;
```

Reply:
Yes, it was my first edition of PR. But test has failed:
```ts
// $ExpectType { id?: string | undefined; s: string | undefined; }
const optionalParams = useParams<OptionalParams>();
```

Reply:
Understandable. Though I'm questioning this test. Do you have an idea why it is important to return 
`{ id?: string | undefined; s: string | undefined; }`
and not

`{ id?: string; s: string | undefined; }`

Reply:
I would go with your initial intuition and change the test. Maybe @pawfa can expand the test a bit more so that it is clear what the intention of #43213 was.

Reply:
Typescript adds `undefined` to optional props.
As I understand the problem is `$ExpectType` requires type as string equality. If you even have extra whitespace or vise versa then test will be failed.

Reply:
Let's ignore `$ExpectType` for this. Why do we need `{ foo?: string | undefined }` instead of `{ foo?: string }`

Reply:
If we return just `Params`, `Params` resolve to `OptionalParams` but `OptionalParams` is not equals to `{ id?: string | undefined; s: string | undefined; }` so the test failes.

Reply:
I understand why it fails. I'm asking why the test/code would need ´id?: string | undefied` instead of `id?: string`? What behavior are we expecting.

Reply:
> Let's ignore $ExpectType for this. Why do we need { foo?: string | undefined } instead of { foo?: string }
They are the same. `{ foo?: string }` is expanded to `{ foo?: string | undefined }` by typescript.

Reply:
So you'd agree that we can just return `Params` and change the test?

Reply:
Let's leave test as is. New changes cover all cases so no problems should arise.

Reply:
> So you'd agree that we can just return Params and change the test?

Completely agree. If you think that important I can change the test.